proto/def.bzl

```diff
@@ -99,21 +99,23 @@ def _go_proto_library_impl(ctx):
         srcs = go_srcs,
     )
     source = go.library_to_source(go, ctx.attr, library, False)
-    if not valid_archive:
-        return [library, source]
-    archive = go.archive(go, source)
-    return [
-        library,
-        source,
-        archive,
-        DefaultInfo(
-            files = depset([archive.data.file]),
-            runfiles = archive.runfiles,
-        ),
-        OutputGroupInfo(
-            compilation_outputs = [archive.data.file],
-        ),
-    ]
+    providers = [library, source]
+    if valid_archive:
+        archive = go.archive(go, source)
+        providers.extend([
+            archive,
+            DefaultInfo(
+                files = depset([archive.data.file]),
+                runfiles = archive.runfiles,
+            ),
+            OutputGroupInfo(
+                compilation_outputs = [archive.data.file],
+            ),
+        ])
+    return struct(
```

Comment:
Why is the legacy provider format needed?

Reply:
We should only migrate to legacy providers if there is a very good reason to do so *temporarily*. I expect they will be removed in the not-too-distant future, and I don't want to do anything to block that removal. What's the plan for after they are removed?

Reply:
I agree with temporarily. As for plan, ideally things happen in this order:
1. This PR is merged (ie go_proto_library exports legacy provider aspect_proto_go_api_info)
2. Bazel adds support for conditional load statements (I've seen this feature mentioned by [[various]](https://github.com/bazelbuild/intellij/issues/214#issuecomment-358765749) [[googlers]](https://github.com/bazelbuild/intellij/issues/214#issuecomment-359012919), but lack detail on what this would look like or when it will happen)
3. `bazelbuild/intellij`'s aspect migrates to the GoSource provider for accumulating generated go proto srcs (which, because it can now _conditionally_ load the GoSource symbol, it does without breaking non-go workspaces).
4. go_proto_library stops exporting aspect_proto_go_api_info (effectively reverting the majority of this PR)
5. Bazel removes legacy providers

Does this seem like a reasonable plan? (pending a compatible timeline for 'conditional load statements')

Reply:
Everything here seems reasonable except step 2. Conditional load statements would be a significant change to Starlark. Without a proposal with some buy-in from the Bazel team, I'd hesitate to approve this as-is.

I've posted a question on the bazel-discuss mailing list to see if there is any way to access new style providers without loading their constructors.

If there isn't, a smaller change would be to add some reflection mechanism (so maybe `target["GoSource"]` would work as well as `target[GoSource]`).

If adding an output group is workable for you though, I'd much prefer that.

Reply:
I think we should explore what a reasonable provider that was both language and editor neutral would look like.
If we can describe that, we can make it a standard feature that will enable better editor support for all languages in all editors.

Reply:
@ianthehat If an output group works (which it sounds like it will below), I think that would be the best option in the short term.

Defining a standard provider, while ideal, would require a lot of coordination across rule sets. Could we worth doing, especially for `x_proto_library` rules though. Maybe it's something we can discuss at the summit in Q1. There's some risk of mismatched versions, and there will always be an oddball rule set that doesn't support it.